### PR TITLE
Handle methods that dispatch to intrinsics on fast path

### PR DESCRIPTION
**Pull request by jez**
### Motivation

Some of our intrinsics dispatch to other methods. Given that we currently use names not symbols to list the downstream files that might need to be retypechecked, it means we have to also record usages of these dispatched-to names as well.

This mostly affects stale errors that could have been occurring around usages of `<call-with-*>` helpers. It could also have affected `initialize`/`new`, but in a large enough codebase and with the changes we made to compute the "max files on fast path" in canTakeFastPath instead of in runFastPath, it was unlikely to still have much effect there (of course, it would matter in small codebases with fewer than 100 files).

### Test plan

See included automated tests.

### CODE REVIEW

File: core/Symbols.h [hunk 1]
```diff
@@ -29,6 +29,18 @@ class SerializerImpl;
...
+    // Used to ensure that calls involving intrinsics are correctly typechecked on the fast path.
+    //
+    // If your intrinsic does not dispatch to another method, the default implementation returns an
+    // empty vector, so you can elide an implementation.
+    //
+    // If your intrinsic cannot declare the list of methods it might dispatch to (like
+    // `<call-with-splat>`, etc.), you will have to edit Substitute.cc.
+    virtual std::vector<NameRef> dispatchesTo() const {
+        return {};
+    }
```
Comment (froydnj) [inline core/Symbols.h:42]: Can we move this out-of-line?

(The first non-abstract virtual method in a class is special, and triggers the output of all kinds of information in the compilation unit which it appears; if that method is defined in a header...the information is output in every file where that header is included.  https://itanium-cxx-abi.github.io/cxx-abi/abi.html#vague-vtable )

File: ast/substitute/Substitute.cc [hunk 2]
```diff
@@ -116,9 +117,26 @@ template <typename T> class SubstWalk {
                     [[maybe_unused]] auto _substituted = subst.substituteSymbolName(name);
                 }
             }
+        } else if (send.fun == core::Names::callWithSplat() || send.fun == core::Names::callWithBlock() ||
+                   send.fun == core::Names::callWithSplatAndBlock() || send.fun == core::Names::checkAndAnd()) {
+            ENFORCE(send.numPosArgs() > 2, "These are special desugar methods, should have at least two args");
+
+            // We're lucky because all of these methods have the symbol they dispatch to as the
+            // positional arg at index 1.
+            auto name = unwrapLiteralToName(send.getPosArg(1));
```
Comment (froydnj) [inline ast/substitute/Substitute.cc:126]: Perhaps we should `ENFORCE` that the arg is actually a literal?  I guess it's basically the same thing as `ENFORCE`'ing that the name exists further down, but it is a little more clear if the assertion triggered.

**Review by froydnj — APPROVED** [review]

File: ast/substitute/Substitute.cc (hunk 2 above)
Comment (jez) [reply ast/substitute/Substitute.cc:126]: Yeah it seemed to me like doing the ENFORCE earlier would have just duplicated the contents of the `unwrapLiteralToName` function. I've added a message to the ENFORCE below to make that hopefully more clear.

File: core/Symbols.h (hunk 1 above)
Comment (jez) [reply core/Symbols.h]: I've re-requested review to make sure I correctly did the thing you were asking for here.

**Review by froydnj — APPROVED** [review]

**Pull request merged**